integration-tests/datasource/src/test/java/io/quarkus/it/datasource/ValidationQueryTest.java

```diff
@@ -0,0 +1,79 @@
+package io.quarkus.it.datasource;
+
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import jakarta.inject.Inject;
+
+import org.h2.jdbc.JdbcSQLNonTransientConnectionException;
+import org.h2.tools.Server;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.agroal.api.AgroalDataSource;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ValidationQueryTest {
+
+    @BeforeEach
+    public void setup() throws SQLException {
+        startH2DB();
+    }
+
+    @AfterEach
+    public void tearDown() {
+        stopH2DB();
+    }
+
+    //tag::injection[]
+    @Inject
+    AgroalDataSource defaultDataSource;
+    //end::injection[]
+
+    private Server tcpServer;
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withConfigurationResource("application-datasource-with-validation.properties");
+
+    @Test
+    public void testConnectionValidation() throws SQLException {
+
+        //Test connection is acquirable
+        Connection connection = defaultDataSource.getConnection();
+        connection.close();
+
+        //Stop DB server
+        stopH2DB();
+
+        //Test connection is NOT available after DB shutdown
+        assertThatThrownBy(() -> {
+            defaultDataSource.getConnection();
+        }).isInstanceOf(JdbcSQLNonTransientConnectionException.class);
+
+        //Start DB server
+        startH2DB();
+
+        //Test connection is acquirable after restart
+        assertThatCode(() -> {
+            defaultDataSource.getConnection();
+        }).doesNotThrowAnyException();
```

Comment:
Do I understand correctly that without the timeout, this assertion would fail (an exception would be thrown)?

Reply:
>  Testing whether a query is issued with a timeout to a JDBC driver is an option, but it requires a mocked JDBC driver.

That's your best bet I think... I see H2 has lots of event listeners, so maybe you can take advantage of that? https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/EventListener.html

Though other PRs I've seen would simply retrieve the Agroal config at runtime and check that it matches what we set up in Quarkus. It's not perfect, but it's reasonable if the alternative is really complex. See for example https://github.com/quarkusio/quarkus/pull/47118/files#diff-b55a3cac0ebfa408f52fff084b639bf946ebcffff8b8b327b7d13373df67c44dR22-R45